torch/fx/experimental/symbolic_shapes.py

```diff
@@ -845,6 +852,12 @@ def render(self):
     def is_equal(self, source1, source2):
         return self._find(source1) == self._find(source2)
 
+    def is_derived(self, src, symbol, expr):
+        return any(
+            src == source and fn(symbol) == expr
+            for source, _root, fn in self.derived_equalities
+        )
```

Comment:
In particular, here, we can potentially have `x = f(y)` and `y = g(z)` and then if we query a guard `x = f(g(z))` we ought to return True when saying that this is_derived.

Reply:
Agreed.

Reply:
Yeah, we now have an algorithm for this. See comments. We normalize everything (e.g., sources use their union-find root) and then express everything in terms of normal Dims and check syntactic equality. Normal dims are equal iff they are the same.